genesis/process/genesisBlockCreator.go

```diff
@@ -0,0 +1,174 @@
+	if check.IfNil(arg.Accounts) {
+		return process.ErrNilAccountsAdapter
+	}
+	if check.IfNil(arg.PubkeyConv) {
+		return process.ErrNilPubkeyConverter
+	}
+	if check.IfNil(arg.InitialNodesSetup) {
+		return process.ErrNilNodesSetup
+	}
+	if check.IfNil(arg.Economics) {
+		return process.ErrNilEconomicsData
+	}
+	if check.IfNil(arg.ShardCoordinator) {
+		return process.ErrNilShardCoordinator
+	}
+	if check.IfNil(arg.Store) {
+		return process.ErrNilStore
+	}
+	if check.IfNil(arg.Blkc) {
+		return process.ErrNilBlockChain
+	}
+	if check.IfNil(arg.Marshalizer) {
+		return process.ErrNilMarshalizer
+	}
+	if check.IfNil(arg.Hasher) {
+		return process.ErrNilHasher
+	}
+	if check.IfNil(arg.Uint64ByteSliceConverter) {
+		return process.ErrNilUint64Converter
+	}
+	if check.IfNil(arg.DataPool) {
+		return process.ErrNilMetaBlocksPool
+	}
+	if check.IfNil(arg.GenesisParser) {
+		return genesis.ErrNilGenesisParser
+	}
+	if arg.GasMap == nil {
+		return process.ErrNilGasSchedule
+	}
+
+	return nil
+}
+
+// CreateGenesisBlocks will try to create the genesis blocks for all shards
+func (gbc *genesisBlockCreator) CreateGenesisBlocks() (map[uint32]data.HeaderHandler, error) {
+	genesisBlocks := make(map[uint32]data.HeaderHandler)
+	var err error
+	var genesisBlock data.HeaderHandler
+	var newArgument ArgsGenesisBlockCreator
+
+	for shardId := uint32(0); shardId < gbc.arg.ShardCoordinator.NumberOfShards(); shardId++ {
+		newArgument, err = gbc.getNewArgForShard(shardId)
+		if err != nil {
+			return nil, fmt.Errorf("'%w' while creating new argument for shard %d",
+				err, shardId)
+		}
+
+		genesisBlock, err = gbc.shardCreatorHandler(newArgument)
+		if err != nil {
+			return nil, fmt.Errorf("'%w' while generating genesis block for shard %d",
+				err, shardId)
+		}
+
+		genesisBlocks[shardId] = genesisBlock
+		err = gbc.saveGenesisBlock(genesisBlock)
+		if err != nil {
+			return nil, fmt.Errorf("'%w' while saving genesis block for shard %d",
+				err, shardId)
+		}
+	}
+
+	newArgument, err = gbc.getNewArgForShard(core.MetachainShardId)
+	if err != nil {
+		return nil, fmt.Errorf("'%w' while creating new argument for metachain", err)
+	}
+
+	newArgument.Blkc = blockchain.NewMetaChain()
+	genesisBlock, err = gbc.metaCreatorHandler(newArgument)
+	if err != nil {
+		return nil, fmt.Errorf("'%w' while generating genesis block for metachain", err)
+	}
+
+	genesisBlocks[core.MetachainShardId] = genesisBlock
+	err = gbc.saveGenesisBlock(genesisBlock)
+	if err != nil {
+		return nil, fmt.Errorf("'%w' while saving genesis block for metachain", err)
+	}
+
+	return genesisBlocks, nil
+}
+
+func (gbc *genesisBlockCreator) getNewArgForShard(shardId uint32) (ArgsGenesisBlockCreator, error) {
+	var err error
+	newArgument := gbc.arg //copy the arguments
+
+	isCurrentShard := shardId == gbc.arg.ShardCoordinator.SelfId()
+	shouldRecreate := !isCurrentShard || newArgument.StartEpochNum != 0
+	if shouldRecreate {
```

Comment:
maybe you can change logic here if !shouldRecreate { return newArguments, err } .... reacreate argument

Reply:
and copy argument just if  shouldRecreate

Reply:
right, is already being copied as I passed it as value